src/Features/LanguageServer/Protocol/Extensions/SourceGeneratedDocumentUris.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer
+{
+    // For source generated documents, we'll produce a URI specifically for LSP that has a scheme the client can register for; the "host" portion will
+    // just be the project ID of the document, and the path will be the hint text for the document. This recognizes that VS Code shows just the local
+    // path portion in the UI and thus embedding more into the path will appear and we don't want that. The rest of the stuff to serialize, namely the DocumentId, and any information
+    // for the SourceGeneratedDocumentIdentity are put in as query string arguments
+    //
+    // For example, the URI can look like:
+    //
+    // roslyn-source-generated://E7D5BCFA-E345-4029-9D12-3EDCD0FB0F6B/Generated.cs?documentId=8E4C0B71-4044-4247-BDD0-04AF4C9E1677&assembly=Generator...
+    //
+    // where the first GUID is the project ID, the second GUID is the document ID.
+    internal static class SourceGeneratedDocumentUris
+    {
+        public const string Scheme = "roslyn-source-generated";
+        private const string GuidFormat = "D";
+
+        public static Uri Create(SourceGeneratedDocumentIdentity identity)
+        {
+            // Ensure the hint path is converted to a URI-friendly format
+            var hintPathParts = identity.HintName.Split('\\');
+            var hintPathPortion = string.Join("/", hintPathParts.Select(Uri.EscapeDataString));
+
+            var uri = Scheme + "://" +
+                identity.DocumentId.ProjectId.Id.ToString(GuidFormat) + "/" +
+                hintPathPortion +
+                "?documentId=" + identity.DocumentId.Id.ToString(GuidFormat) +
+                "&hintName=" + Uri.EscapeDataString(identity.HintName) +
+                "&assemblyName=" + Uri.EscapeDataString(identity.Generator.AssemblyName) +
+                "&assemblyVersion=" + Uri.EscapeDataString(identity.Generator.AssemblyVersion.ToString()) +
+                "&typeName=" + Uri.EscapeDataString(identity.Generator.TypeName);
+
+            // If we have a path (which is technically optional) also append it
+            if (identity.Generator.AssemblyPath != null)
+                uri += "&assemblyPath=" + Uri.EscapeDataString(identity.Generator.AssemblyPath);
```

Comment:
Am slightly concerned about the size of the URI - wondering if having all this data present in the URI itself could hit serialization perf.  

Is all of this strictly necessary to include, or could we infer it back on the server side based on the document id?

Reply:
do we need this assmebly path?

Reply:
> Is all of this strictly necessary to include, or could we infer it back on the server side based on the document id?

Initially I was just doing document ID; the problems began once we actually were freezing the source generated documents to match the contents of the text changes from LSP. There we use the identity as a general concept for matching up the source generated document data at the workspace layer, because the underlying workspace APIs allow a frozen generated document to be created "out of thin air" if we realized asynchronously that the generated document doesn't exist anymore.

We could reduce this back to document ID and hold onto the identity on the server side for open source generated documents as a separate map, but it gets a bit funky that way -- when do we put something into that map? Unlike VS for Windows when it's a very clear "here's a generated document, now we know we're navigating to it" gesture, here we're just returning a URI and we might get a later request for it. Or maybe we won't. There's no guarantee so just packing the data in the URI seemed the easiest way to do it.

I expect the serialization cost to only be paid any time a user command like go to def or find refs touches a generated document. In the former case, the cost is trivial; the latter case could be expensive. But I'd rather think we should measure the cost first to understand if there's an easier change; for example if the expensive bit is the Uri.EscapeDataString() then we could easily stick that bit in a cache.

Reply:
> do we need this assmebly path?

It's used to differentiate different generators having the same assembly name. We don't support that per se but we've had fun bugs when users (even accidentally) did that. Many gold bars, many complaints. So at this point we just carry this around too.

Reply:
I think the main area that large URIs would bite us is something like completion or code lens - the data objects there often contain URIs, which if they get large can be quite expensive.

Completion I don't think is as much of an issue - these are readonly documents so completion shouldn't be active.  Code lens would be active though, so we'd be serializing this for each code lens item in the document.

I guess we can try and see if we have issues, but generally the less data we can put in there the better.

Reply:
> these are readonly documents so completion shouldn't be active

Hi, it's me, Razor.

Reply:
> I guess we can try and see if we have issues, but generally the less data we can put in there the better.

This was generally my approach: this is "correct" in that it accurately represents the underlying Roslyn model. If it's slow but the only reason it's slow is because of the code creating the string, we can easily add a cache. If it's slow because just having a big string is slow even if it's instant to make, then we can revisit.

And in the completion request I imagine there might be one bigger URI for the request, but the response is still massive, right? Or we even seeing places where the _request_ is too big?

Reply:
Ah right - if Razor is querying the c# source generated docs we'd put it in the data.  But now I'm thinking about it more, I believe we promote the data onto the list instead of each item (to avoid this with existing URIs)

Reply:
I should be clear here, Razor is not currently querying source generated docs via LSP, but it might depending on order of cohost-y things.